File: core-py/src/circle_agent_kit/contracts.py
"""SDK-native contract execution, CCTP bridge, and swap.

Mirrors packages/core-ts/src/contracts.ts. Built on the developer-controlled
wallet the rest of the kit uses — no Circle CLI or separate session required.
"""

from __future__ import annotations

import time
import urllib.parse
import urllib.request
import json
import uuid
from typing import Any, Optional

from .chains import CCTP_V2, get_chain
from .errors import err
from .requests import to_base_units


def address_to_bytes32(address: str) -> str:
    clean = (address or "").replace("0x", "").lower()
    if len(clean) != 40:
        raise err("VALIDATION", f'Invalid EVM address: "{address}".')
    return "0x" + ("0" * 24) + clean


def _http_get_json(url: str) -> Optional[dict]:
    try:
        with urllib.request.urlopen(url, timeout=30) as resp:  # noqa: S310
            if resp.status != 200:
                return None
            return json.loads(resp.read().decode("utf-8"))
    except Exception:
        return None


def _http_get_quote(url: str, headers: dict) -> dict:
    req = urllib.request.Request(url, headers=headers)  # noqa: S310
    with urllib.request.urlopen(req, timeout=30) as resp:  # noqa: S310
        body = resp.read().decode("utf-8")
        if resp.status != 200:
            raise err("UPSTREAM", f"Swap quote failed ({resp.status}): {body}")
        return json.loads(body)


def get_bridge_attestation(
    config,
    source_chain: str,
    burn_tx_hash: str,
    interval_s: float = 5.0,
    timeout_s: float = 300.0,
) -> dict:
    domain = get_chain(source_chain).cctp_domain
    if domain is None:
        raise err("VALIDATION", f"Chain {source_chain} has no CCTP domain.")
    base = (
        CCTP_V2["attestation_api"]["mainnet"]
        if config.network == "MAINNET"
        else CCTP_V2["attestation_api"]["testnet"]
    )
    url = f"{base}/v2/messages/{domain}?transactionHash={burn_tx_hash}"
    start = time.time()
    while True:
        body = _http_get_json(url)
        if body:
            msg = (body.get("messages") or [{}])[0]
            if msg.get("status") == "complete" and msg.get("attestation") and msg.get("message"):
                return {
                    "status": "complete",
                    "message": msg["message"],
                    "attestation": msg["attestation"],
                }
        if time.time() - start > timeout_s:
            raise err(
                "TRANSACTION_FAILED",
                f"Timed out waiting for CCTP attestation of {burn_tx_hash}.",
            )
        time.sleep(interval_s)


def get_swap_quote(
    config,
    chain: str,
    sell_token: str,
    buy_token: str,
    sell_amount: str,
    taker_address: str,
    slippage_bps: Optional[int] = None,
) -> dict:
    if not config.swap_api_url:
        raise err(
            "CONFIG_MISSING",
            "Swap requires a DEX aggregator. Set SWAP_API_URL (0x-compatible) and SWAP_API_KEY, "
            "or use the Agent Stack CLI swap (agent_swap) which uses Circle's hosted router.",
        )
    chain_id = get_chain(chain).chain_id
    params = {
        "chainId": str(chain_id or ""),
        "sellToken": sell_token,
        "buyToken": buy_token,
        "sellAmount": sell_amount,
        "taker": taker_address,
    }
    if slippage_bps is not None:
        params["slippageBps"] = str(slippage_bps)
    qs = urllib.parse.urlencode(params)
    headers = {"0x-version": "v2"}
    if config.swap_api_key:
        headers["0x-api-key"] = config.swap_api_key
    q = _http_get_quote(f"{config.swap_api_url}/swap/allowance-holder/quote?{qs}", headers)
    tx = q.get("transaction") or q
    return {
        "sellToken": sell_token,
        "buyToken": buy_token,
        "sellAmount": sell_amount,
        "buyAmount": q.get("buyAmount"),
        "allowanceTarget": q.get("allowanceTarget")
        or (q.get("issues", {}).get("allowance") or {}).get("spender")
        or tx.get("to"),
        "to": tx.get("to"),
        "data": tx.get("data"),
        "value": tx.get("value") or "0",
    }

File: core-ts/src/contracts.ts
import { v4 as uuidv4 } from "uuid";
import type { CircleWalletsClient } from "./client.js";
import type { CircleAgentConfig } from "./config.js";
import { getChain, ID_TO_APPKIT_CHAIN } from "./chains.js";
import { err } from "./errors.js";
import { getAddress } from "./wallets.js";
import type { TransactionInfo } from "./types.js";

/**
 * SDK-native DeFi capabilities that the Circle CLI Agent Stack exposes but the
 * developer-controlled-wallets SDK doesn't offer as one-liners:
 *
 *  - executeContract: arbitrary contract write (createContractExecutionTransaction)
 *  - bridgeUSDC:      CCTP v2 (approve -> depositForBurn -> attest -> receiveMessage)
 *  - swap:            via a configured DEX aggregator (quote -> approve -> execute)
 *
 * All are built on the same wallet the rest of the kit uses, so no CLI or
 * separate session is required.
 */

const feeConfig = (feeLevel: "LOW" | "MEDIUM" | "HIGH" = "MEDIUM") => ({
  type: "level",
  config: { feeLevel },
});

/** Left-pad an EVM address to a 32-byte hex string (CCTP bytes32 encoding). */
export function addressToBytes32(address: string): string {
  const clean = address.replace(/^0x/, "").toLowerCase();
  if (clean.length !== 40) throw err("VALIDATION", `Invalid EVM address: "${address}".`);
  return "0x" + "0".repeat(24) + clean;
}

function appKitChain(id: string): any {
  const mapped = ID_TO_APPKIT_CHAIN[id.toUpperCase()];
  if (!mapped) throw err("VALIDATION", `Chain "${id}" is not supported by App Kit.`);
  return mapped;
}

/** Execute a write function on a contract from a wallet. */
export async function executeContract(
  client: CircleWalletsClient,
  params: {
    walletId: string;
    contractAddress: string;
    abiFunctionSignature?: string;
    abiParameters?: unknown[];
    callData?: string;
    amount?: string;
    feeLevel?: "LOW" | "MEDIUM" | "HIGH";
  }
): Promise<TransactionInfo> {
  if (!params.abiFunctionSignature && !params.callData) {
    throw err("VALIDATION", "executeContract needs either abiFunctionSignature or callData.");
  }
  const res = await client.createContractExecutionTransaction({
    walletId: params.walletId,
    contractAddress: params.contractAddress,
    abiFunctionSignature: params.abiFunctionSignature,
    abiParameters: params.abiParameters,
    callData: params.callData,
    amount: params.amount,
    fee: feeConfig(params.feeLevel),
    idempotencyKey: uuidv4(),
  });
  const id = res?.data?.id;
  if (!id) throw err("UPSTREAM", "Contract execution returned no transaction id.", res);
  return { id, state: res?.data?.state ?? "INITIATED" };
}

export interface BridgeResult {
  fromChain: string;
  toChain: string;
  amount: string;
  burnTxId: string;
  burnTxHash?: string;
  attestation?: string;
  mintTxId?: string;
  mintTxHash?: string;
  state: "BURNED" | "COMPLETE";
}

/**
 * Bridge USDC across chains via CCTP v2. Uses the Circle App Kit to handle
 * the complete flow (approve -> burn -> attest -> mint).
 */
export async function bridgeUSDC(
  client: CircleWalletsClient,
  _config: CircleAgentConfig,
  params: {
    fromChain: string;
    toChain: string;
    sourceWalletId: string;
    destWalletId: string;
    amount: string;
    mintRecipient?: string;
    waitForMint?: boolean;
  }
): Promise<BridgeResult> {
  const sourceAddress = await getAddress(client, params.sourceWalletId);
  const destAddress = params.mintRecipient ?? (await getAddress(client, params.destWalletId));

  const result = await client.appKit.bridge({
    from: {
      adapter: client.adapter,
      chain: appKitChain(params.fromChain),
      address: sourceAddress,
    },
    to: {
      adapter: client.adapter,
      chain: appKitChain(params.toChain),
      address: destAddress,
    },
    amount: params.amount,
  });

  const burnStep = result.steps.find((s: any) => s.name.toLowerCase() === "burn");
  const mintStep = result.steps.find((s: any) => s.name.toLowerCase() === "mint");

  return {
    fromChain: params.fromChain,
    toChain: params.toChain,
    amount: params.amount,
    burnTxId: burnStep?.txHash ?? "pending",
    burnTxHash: burnStep?.txHash,
    mintTxId: mintStep?.txHash,
    mintTxHash: mintStep?.txHash,
    state: result.state === "success" ? "COMPLETE" : "BURNED",
  };
}

export interface SwapQuote {
  sellToken: string;
  buyToken: string;
  sellAmount: string;
  buyAmount: string;
  /** Router/spender to approve. */
  allowanceTarget: string;
  /** Router contract to call. */
  to: string;
  /** Raw calldata for the swap. */
  data: string;
  /** Native value to send (usually "0" for ERC20->ERC20). */
  value?: string;
}

/**
 * Fetch a swap quote from the Circle App Kit.
 */
export async function getSwapQuote(
  client: CircleWalletsClient,
  config: CircleAgentConfig,
  params: {
    chain: string;
    sellToken: string;
    buyToken: string;
    sellAmount: string; // base units
    takerAddress: string;
    slippageBps?: number;
  }
): Promise<SwapQuote> {
  const estimate = await client.appKit.estimateSwap({
    from: {
      adapter: client.adapter,
      chain: appKitChain(params.chain),
      address: params.takerAddress,
    },
    tokenIn: params.sellToken,
    tokenOut: params.buyToken,
    amountIn: params.sellAmount,
    config: {
      kitKey: config.kitKey,
      slippageBps: params.slippageBps,
    },
  });

  return {
    sellToken: params.sellToken,
    buyToken: params.buyToken,
    sellAmount: params.sellAmount,
    buyAmount: estimate.estimatedOutput.amount,
    allowanceTarget: "", // App Kit handles approval internally
    to: "",
    data: "",
  };
}

/**
 * Execute a token swap from a wallet using the Circle App Kit.
 */
export async function swap(
  client: CircleWalletsClient,
  config: CircleAgentConfig,
  params: {
    walletId: string;
    walletAddress: string;
    chain: string;
    sellToken: string;
    buyToken: string;
    sellAmount: string; // base units
    slippageBps?: number;
    feeLevel?: "LOW" | "MEDIUM" | "HIGH";
  }
): Promise<{ quote: any; swapTxId: string }> {
  const result = await client.appKit.swap({
    from: {
      adapter: client.adapter,
      chain: appKitChain(params.chain),
      address: params.walletAddress,
    },
    tokenIn: params.sellToken,
    tokenOut: params.buyToken,
    amountIn: params.sellAmount,
    config: {
      kitKey: config.kitKey,
      slippageBps: params.slippageBps,
    },
  });

  return {
    quote: {
      sellToken: params.sellToken,
      buyToken: params.buyToken,
      sellAmount: params.sellAmount,
      buyAmount: result.amountOut,
    },
    swapTxId: result.txHash,
  };
}
